### PR TITLE
refactor: use interface references in public api (awslint:ref-via-interface)

### LINKED ISSUE #2412: awslint:ref-via-interface
All public APIs (props, methods, properties) should only use construct interfaces (IFoo)

### PR DESCRIPTION
**Pull request by eladb**
Adds a new awslint:ref-via-interface rule which validates that all input APIs
(e.g. props, method arguments) use construct interface (IBucket) and not concrete
classes (Bucket). This is in order to enable passing in unowned resources and
in accordance with the aws construct library guidelines.

There are situations where an owned resource is required. In those, the rule
can be disabled by adding [disable-awslint:ref-via-interface] to the element's
inline documentation.

To enable this, the following new construct interfaces were added, along with `fromXxx` import methods:
* `applicationautoscaling.IScalableTarget`
* `cloudwatch.IAlarm`
* `ecs.IService`
* `ecs.IEc2Service`
* `ec2.IFargateService`
* `ecs.ITaskDefinition`
* `iam.IGroup`
* `iam.IUser`
* `iam.IPolicy`
* `lambda.IVersion`

Fixes #2412 

BREAKING CHANGE: `apigateway.ResourceBase.trackChild` is now internal.
* `cloudfront.S3OriginConfig.originAccessIdentity` is now `originAccessIdentityId`
* `codedeploy.LambdaDeploymentGroup.alarms` is now `cloudwatch.IAlarm[]` (previously `cloudwatch.Alarm[]`)
* `codepipeline.crossRegionScaffoldingStacks` renamed to `crossRegionScaffolding`
* `codepipeline.CrossRegionScaffoldingStack` renamed to `codepipeline.CrossRegionScaffolding` and cannot be instantiated (abstract)
* `ec2.VpcSubnet.addDefaultRouteToNAT` renamed to `addDefaultNatRoute` and made public
* `ec2.VpcSubnet.addDefaultRouteToIGW` renamed to `addDefaultInternetRoute`, made public and first argument is the gateway ID (string) and not the CFN L1 class
* `ecs.Ec2EventRuleTarget.taskDefinition` is now `ITaskDefinition` (previously `TaskDefinition`)
* `lambda.IEventSource.bind` now accepts `IFunction` instead of `FunctionBase`. Use `IFunction.addEventSourceMapping` to add an event source mapping under the function.
* `lambda.Layer.grantUsage` renamed to `lambda.layer.addPermission` and returns void
* `stepfunctions.StateMachine.role` is now `iam.IRole` (previously `iam.Role`)

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: tools/awslint/lib/rules/api.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      if (type.type && type.type.isInterfaceType()) {
+        return assertInterface(type.type);
+      }
+
+      // enums are okay
+      if (type.type && type.type.isEnumType()) {
+        return;
+      }
+
+      // classes are okay as long as they are not constructs
```
Comment (rix0rrr) [inline tools/awslint/lib/rules/api.ts]: ...resources?

File: packages/@aws-cdk/aws-lambda/lib/layers.ts [hunk 2]
```diff
@@ -54,15 +54,17 @@ export interface ILayerVersion extends IResource {
...
+   * require calling this method. Note that the principal that creates the
+   * Lambda function using the layer (for example, a CloudFormation changeset
+   * execution role) also needs to have the ``lambda:GetLayerVersion``
+   * permission on the layer version.
    *
    * @param id the ID of the grant in the construct tree.
    * @param grantee the identification of the grantee.
    */
-  grantUsage(id: string, grantee: LayerVersionUsageGrantee): ILayerVersion
+  addPermission(id: string, grantee: LayerVersionPermission): void;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-lambda/lib/layers.ts]: Parameter name and type name no longer agree.

File: packages/@aws-cdk/aws-lambda/lib/lambda-version.ts [hunk 3]
```diff
@@ -28,6 +41,11 @@ export interface VersionProps {
   readonly lambda: IFunction;
 }
 
+export interface VersionAttributes {
+  readonly version: string;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-lambda/lib/lambda-version.ts]: Docstrings

File: tools/awslint/lib/rules/api.ts (hunk 1 above)
Comment (eladb) [reply tools/awslint/lib/rules/api.ts]: fixed

File: packages/@aws-cdk/aws-lambda/lib/layers.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-lambda/lib/layers.ts]: renamed to "permission"

File: packages/@aws-cdk/aws-lambda/lib/lambda-version.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/aws-lambda/lib/lambda-version.ts]: added. we'll do another docstrings pass later.

**Pull request merged**